scipy/sparse/coo.py

```diff
@@ -380,16 +380,22 @@ def todok(self):
         return dok
 
     def diagonal(self):
-        # Could be rewritten without the python loop.
-        # Data entries at the same (row, col) are summed.
-        n = min(self.shape)
-        ndata = self.data.shape[0]
-        d = np.zeros(n, dtype=self.dtype)
-        for i in xrange(ndata):
-            r = self.row[i]
-            if r == self.col[i]:
-                d[r] += self.data[i]
-        return d
+        diag_mask = self.row == self.col
+
+        # While it's tempting to use the following fast path when
+        # self.has_canonical_format, timing with random matrices shows that
+        # the speedup is marginal (even when the matrix is diagonal).
+        # diag[self.row[diag_mask]] = self.data[diag_mask]
```

Comment:
We're already testing both canonical and non-canonical sparse matrices in the test suite, so no additional tests would be required.

See the existing coverage for a similar condition here: https://codecov.io/github/scipy/scipy/scipy/sparse/compressed.py?ref=a68a1589e4d0285a32de738f10fab974b83016a7#l-993